core/search/sort.hpp

```diff
@@ -451,153 +280,239 @@ class sort {
+        score_size_(buckets_.size() * sizeof(score_t)),
+        stats_size_{stats_size},
+        features_{features} {}
+
+  OrderBuckets buckets_;
+  size_t score_size_{};
+  size_t stats_size_{};
+  IndexFeatures features_{IndexFeatures::NONE};
+};
+
+struct NoopAggregator {
+  constexpr size_t size() const noexcept { return 0; }
+};
+
+template<typename Merger, size_t Size>
+struct Aggregator : Merger {
+  static_assert(Size > 0);
+
+  constexpr size_t size() const noexcept { return Size; }
+
+  constexpr size_t byte_size() const noexcept { return Size * sizeof(score_t); }
+
+  constexpr score_t* temp() noexcept { return buf.data(); }
+
+  void operator()(score_t* dst, const score_t* src) const noexcept {
+    for (size_t i = 0; i < Size; ++i) {
+      Merger::operator()(i, dst, src);
+    }
   }
 
-  FORCE_INLINE static score_type& score_cast(byte_type* buf) noexcept {
-    return const_cast<score_type&>(score_cast(const_cast<const byte_type*>(buf)));
+  void operator()(score_t* dst, const score_t** src,
+                  size_t size) const noexcept {
+    for (size_t i = 0; i < Size; ++i) {
+      Merger::operator()(i, dst, src, size);
+    }
+  }
+
+  std::array<score_t, Size> buf;
+};
+
+template<typename Merger>
+struct Aggregator<Merger, std::numeric_limits<size_t>::max()> : Merger {
+  explicit Aggregator(size_t size) noexcept : count{size} {
+    assert(size);
+    buf.resize(byte_size());
+  }
+
+  size_t size() const noexcept { return count; }
+
+  size_t byte_size() const noexcept { return size() * sizeof(score_t); }
+
+  score_t* temp() noexcept { return reinterpret_cast<score_t*>(buf.data()); }
+
+  void operator()(score_t* dst, const score_t* src) const noexcept {
+    for (size_t i = 0; i < count; ++i) {
+      Merger::operator()(i, dst, src);
+    }
+  }
+
+  void operator()(score_t* dst, const score_t** src,
+                  size_t size) const noexcept {
+    for (size_t i = 0; i < size; ++i) {
+      Merger::operator()(i, dst, src, size);
+    }
+  }
+
+  size_t count;
+  bstring buf;
+};
+
+template<typename Aggregator>
+constexpr bool HasScore() noexcept {
+  return !std::is_same_v<Aggregator, NoopAggregator>;
+}
+
+struct SumMerger {
+  void operator()(size_t idx, score_t* RESTRICT dst,
+                  const score_t* RESTRICT src) const noexcept {
+    dst[idx] += src[idx];
   }
 
-  static void bulk_aggregate(const order_bucket* ctx, byte_type* dst,
-                             const byte_type** src_begin, size_t size) noexcept {
-    const auto offset = ctx->score_offset;
-    auto& casted_dst = score_cast(dst + offset);
-    casted_dst = ScoreType();
+  void operator()(size_t idx, score_t* dst, const score_t** src_begin,
+                  size_t size) const noexcept {
+    auto& casted_dst = dst[idx];
+    casted_dst = {};
 
     const auto** src_end = src_begin + size;
     const auto** src_next = src_begin + 4;
     for (; src_next <= src_end; src_begin = src_next, src_next += 4) {
```

Comment:
Is manually unroll better than common code here?
Maybe we need use sse/avx manually in that case?